src/catalog/src/durable/objects/state_update.rs

```diff
@@ -288,16 +290,58 @@ impl StateUpdateKindJson {
         StateUpdateKindJson(row)
     }
 
-    pub(crate) fn to_serde<D: serde::de::DeserializeOwned>(self) -> D {
+    pub(crate) fn to_serde<D: serde::de::DeserializeOwned>(&self) -> D {
         self.try_to_serde().expect("jsonb should roundtrip")
     }
 
     pub(crate) fn try_to_serde<D: serde::de::DeserializeOwned>(
-        self,
+        &self,
     ) -> Result<D, serde_json::error::Error> {
         let serde_value = self.0.as_ref().to_serde_json();
         serde_json::from_value::<D>(serde_value)
     }
+
+    fn kind(&self) -> &str {
+        let row = self.0.row();
+        let mut iter = row.unpack_first().unwrap_map().iter();
+        let datum = iter
+            .find_map(|(field, datum)| if field == "kind" { Some(datum) } else { None })
+            .expect("kind field must exist");
+        datum.unwrap_str()
+    }
+
+    /// Returns true if this is an update kind that is always deserializable, even before migrations. Otherwise, returns false.
+    pub(crate) fn is_always_deserializable(&self) -> bool {
+        // Construct some fake update kinds so we can extract exactly what the kind field will
+        // serialize as.
+        static DESERIALIZABLE_KINDS: LazyLock<HashSet<String>> = LazyLock::new(|| {
+            [
+                StateUpdateKind::FenceToken(FenceToken {
+                    deploy_generation: 1,
+                    epoch: Epoch::new(1).expect("non-zero"),
+                }),
+                StateUpdateKind::Config(
+                    proto::ConfigKey { key: String::new() },
+                    proto::ConfigValue { value: 1 },
+                ),
+                StateUpdateKind::Setting(
+                    proto::SettingKey {
+                        name: String::new(),
+                    },
+                    proto::SettingValue {
+                        value: String::new(),
+                    },
+                ),
+            ]
+            .into_iter()
+            .map(|kind| {
+                let json_kind: StateUpdateKindJson = kind.into();
+                json_kind.kind().to_string()
+            })
+            .collect()
+        });
```

Comment:
@ParkMyCar, just to confirm my understanding, this will only ever be created once and be re-used across all method invocations?

Reply:
Yes exactly!